### DISCUSSION TITLE
Why we won't support `styled-components` syntax

### DISCUSSION THREAD

**Discussion by frontsideair** [facebook/stylex#332] [Q&A] [answered]
Any chance of supporting this kind of API? This can be done in userland too, and I may author one myself if it's not considered.

```tsx
export const Button = stylex.styled("button", {
  intent: {
    primary: {
      backgroundColor: colors.primary,
      color: colors.text.body,
    },
    secondary: {
      backgroundColor: colors.secondary,
      color: colors.text.body,
    },
  }
});
```

Some advantages:
- Opaque, can only be styled with variants
- Automatic wiring up of variants
- Automatic ref forwarding
- Zero cruft
- Doesn't change anything in stylex

**Comment by necolas** [comment, marked as answer]
We don't plan to support the limiting "styled-components" kind of syntax. Variants are easy to support without adding a custom API for it. We have an example in the docs: https://stylexjs.com/docs/learn/styling-ui/using-styles/#style-variants

**Reply by frontsideair** [reply]
Love that variants are supported without any new API! And I also agree that template string syntax is very limiting, but my suggestion specifically is using the object syntax to provide a lightweight wrapper that forwards ref, threads props and makes it terse. What do you consider limiting about this proposal when used with object syntax?

Since it's possible to to do this in userland I'll just go ahead and implement it myself, I only wanted to see this supported first-party for adoption and easy onboarding for people who are coming from styled-components.

**Reply by nmn** [reply]
> What do you consider limiting about this proposal when used with object syntax?

We want to avoid increasing the API surface area. 
- There is already a way to define styles that is more flexible. 
- Ref forwarding isn't needed when you use `<button>` directly, so that is not a value-add.
- It adds an additional custom component to the component tree which is overhead
- It hides what is happening, and how props work isn't clear

> Since it's possible to to do this in userland I'll just go ahead and implement it myself,

FWIW, this is not possible to do in user-land unless you write a custom Babel plugin. Feel free to write a Babel plugin to do this.

We're currently focused on fixing any rough edges with the core of StyleX and improving bundler integrations. In the future, we would propose a code-mod to help people coming from `styled-components` to adopt stylex.

**Reply by frontsideair** [reply]
I believe it would decrease the impact to the API surface if this was a separate package in the same org, but it's still something to maintain.

> Ref forwarding isn't needed when you use <button> directly, so that is not a value-add.

My example was bad, so let's assume that it's not a plain `<button>` but a `<ReactAriaButton>`.

> It adds an additional custom component to the component tree which is overhead

If someone was creating a design system, which is the point in my example of creating a reusable component with variants, there would be a component one way or another.

And I missed that a userland solution would miss the collection of styles by the compiler, so a middle ground would be to define variants with stylex like and use them like:

```jsx
const rabStyles = stylex.create({
  intent: {
    primary: {
      backgroundColor: colors.primary,
      color: colors.text.body,
    },
    secondary: {
      backgroundColor: colors.secondary,
      color: colors.text.body,
    },
  }
});

export const Button = stylexStyledAddon(ReactAriaButton, rabStyles);
```

About the codemod, it would definitely ease the pain, but I think users of `styled-components` and `emotion` enjoy writing styles that way so a codemod would only partly helpful for adoption.

In any case I respect the decision, and if my revised example works I'll use it that way.

**Reply by nmn** [reply]
> if this was a separate package in the same org

We wouldn't want to support something like this in the same org. Just having an additional option adds confusion and overhead and we want to keep the API as minimal as possible. We're OK with it being a third-party package though.

> My example was bad, so let's assume that it's not a plain `<button>` but a `<ReactAriaButton>`.

In such an example `ReactAriaButton` would need to support accepting a ref, so the same thing holds. The "ref forwarding" done by `styled-components`/`emotion` is a solution to a problem that only exists because an extra component layer is being added.

>  a middle ground would be to define variants with stylex like and use them like

This is perfectly fine. Feel free to use it if you find it useful, but using styles with `stylex.props` directly on JSX elements will always give you more compile-time optimisation. There is a typo in your example though:

```tsx
const rabStyles = stylex.create({
    primary: {
      backgroundColor: colors.primary,
      color: colors.text.body,
    },
    secondary: {
      backgroundColor: colors.secondary,
      color: colors.text.body,
    },
});
```

**Reply by necolas** [reply]
Also, ref forwarding in particular is something that teams are working on solving more generally for React. We won't need to solve those kinds of ergonomic issues in StyleX.
